### ISSUE TITLE
'.txt' files unsupported in Biologic and Maccor parsing engines

### ISSUE THREAD

**Issue by CWestICL** [ImperialCollegeLondon/Faraday-liionsden#185]
**Describe the issue**
Uploading the example files with the '.txt' extension resulted in UnsupportedFileTypeError for both Biologic and Maccor parsing engines.

**To Reproduce**
Steps to reproduce the behaviour:
1. Uploaded MPG-205+-+163.1.6.154_Group+4+-+Cell+12_01_BCD_C02.mps.txt using the Biologic parser
2. Got an UnsupportedFileTypeError: Error tokenizing data. C error: Expected 1 fields in line 12, saw 2
3. Uploaded TPG1+-+Cell+15+-+002.txt using the Maccor parser
4. Got an UnsupportedFileTypeError: Unknown file extension for Maccor parsing engine '.txt'.

**Tracebacks**
MPG-205+-+163.1.6.154_Group+4+-+Cell+12_01_BCD_C02.mps.txt using the Biologic parser:
```
Environment:


Request Method: POST
Request URL: http://localhost:8000/battDB/exps/add_data/8/

Django Version: 4.1
Python Version: 3.8.14
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'guardian',
 'bootstrap5',
 'django_better_admin_arrayfield',
 'django_tables2',
 'crispy_forms',
 'crispy_bootstrap5',
 'django_filters',
 'django_cleanup.apps.CleanupConfig',
 'rest_framework',
 'rest_framework.authtoken',
 'storages',
 'django_extensions',
 'mptt',
 'battDB.apps.BattdbConfig',
 'common.apps.CommonConfig',
 'dfndb.apps.DfndbConfig',
 'management.apps.ManagementConfig']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']



Traceback (most recent call last):
  File "/usr/src/app/parsing_engines/biologic_engine.py", line 124, in load_biologic_data
    data = pd.read_csv(file_obj, delimiter=",", **kwargs)
  File "/usr/local/lib/python3.8/site-packages/pandas/util/_decorators.py", line 311, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/pandas/io/parsers/readers.py", line 678, in read_csv
    return _read(filepath_or_buffer, kwds)
  File "/usr/local/lib/python3.8/site-packages/pandas/io/parsers/readers.py", line 581, in _read
    return parser.read(nrows)
  File "/usr/local/lib/python3.8/site-packages/pandas/io/parsers/readers.py", line 1253, in read
    index, columns, col_dict = self._engine.read(nrows)
  File "/usr/local/lib/python3.8/site-packages/pandas/io/parsers/c_parser_wrapper.py", line 225, in read
    chunks = self._reader.read_low_memory(nrows)
  File "pandas/_libs/parsers.pyx", line 805, in pandas._libs.parsers.TextReader.read_low_memory
    <source code not available>
  File "pandas/_libs/parsers.pyx", line 861, in pandas._libs.parsers.TextReader._read_rows
    <source code not available>
  File "pandas/_libs/parsers.pyx", line 847, in pandas._libs.parsers.TextReader._tokenize_rows
    <source code not available>
  File "pandas/_libs/parsers.pyx", line 1960, in pandas._libs.parsers.raise_parser_error
    <source code not available>

During handling of the above exception (Error tokenizing data. C error: Expected 1 fields in line 10, saw 2
), another exception occurred:
  File "/usr/src/app/parsing_engines/biologic_engine.py", line 128, in load_biologic_data
    data = pd.read_csv(file_obj, delimiter="\t", **kwargs)
  File "/usr/local/lib/python3.8/site-packages/pandas/util/_decorators.py", line 311, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/pandas/io/parsers/readers.py", line 678, in read_csv
    return _read(filepath_or_buffer, kwds)
  File "/usr/local/lib/python3.8/site-packages/pandas/io/parsers/readers.py", line 581, in _read
    return parser.read(nrows)
  File "/usr/local/lib/python3.8/site-packages/pandas/io/parsers/readers.py", line 1253, in read
    index, columns, col_dict = self._engine.read(nrows)
  File "/usr/local/lib/python3.8/site-packages/pandas/io/parsers/c_parser_wrapper.py", line 225, in read
    chunks = self._reader.read_low_memory(nrows)
  File "pandas/_libs/parsers.pyx", line 805, in pandas._libs.parsers.TextReader.read_low_memory
    <source code not available>
  File "pandas/_libs/parsers.pyx", line 861, in pandas._libs.parsers.TextReader._read_rows
    <source code not available>
  File "pandas/_libs/parsers.pyx", line 847, in pandas._libs.parsers.TextReader._tokenize_rows
    <source code not available>
  File "pandas/_libs/parsers.pyx", line 1960, in pandas._libs.parsers.raise_parser_error
    <source code not available>

During handling of the above exception (Error tokenizing data. C error: Expected 1 fields in line 12, saw 2
), another exception occurred:
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/views/generic/base.py", line 103, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/guardian/mixins.py", line 213, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/views/generic/base.py", line 142, in dispatch
    return handler(request, *args, **kwargs)
  File "/usr/src/app/battDB/views.py", line 162, in post
    form.instance.full_clean()
  File "/usr/local/lib/python3.8/site-packages/django/db/models/base.py", line 1471, in full_clean
    self.clean()
  File "/usr/src/app/battDB/models.py", line 578, in clean
    parsed_file = parse_data_file(file_obj, file_format, columns=cols)
  File "/usr/src/app/parsing_engines/parsing_engines_base.py", line 241, in parse_data_file
    engine = get_parsing_engine(file_format).factory(file_obj)
  File "/usr/src/app/parsing_engines/biologic_engine.py", line 48, in factory
    data = load_biologic_data(file_obj, skip_rows, cls.encoding)
  File "/usr/src/app/parsing_engines/biologic_engine.py", line 130, in load_biologic_data
    raise UnsupportedFileTypeError(err)

Exception Type: UnsupportedFileTypeError at /battDB/exps/add_data/8/
Exception Value: Error tokenizing data. C error: Expected 1 fields in line 12, saw 2


```

TPG1+-+Cell+15+-+002.txt using the Maccor parser:
```
Environment:


Request Method: POST
Request URL: http://localhost:8000/battDB/exps/add_data/13/

Django Version: 4.1
Python Version: 3.8.14
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'guardian',
 'bootstrap5',
 'django_better_admin_arrayfield',
 'django_tables2',
 'crispy_forms',
 'crispy_bootstrap5',
 'django_filters',
 'django_cleanup.apps.CleanupConfig',
 'rest_framework',
 'rest_framework.authtoken',
 'storages',
 'django_extensions',
 'mptt',
 'battDB.apps.BattdbConfig',
 'common.apps.CommonConfig',
 'dfndb.apps.DfndbConfig',
 'management.apps.ManagementConfig']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']



Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/views/generic/base.py", line 103, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/guardian/mixins.py", line 213, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/django/views/generic/base.py", line 142, in dispatch
    return handler(request, *args, **kwargs)
  File "/usr/src/app/battDB/views.py", line 162, in post
    form.instance.full_clean()
  File "/usr/local/lib/python3.8/site-packages/django/db/models/base.py", line 1471, in full_clean
    self.clean()
  File "/usr/src/app/battDB/models.py", line 578, in clean
    parsed_file = parse_data_file(file_obj, file_format, columns=cols)
  File "/usr/src/app/parsing_engines/parsing_engines_base.py", line 241, in parse_data_file
    engine = get_parsing_engine(file_format).factory(file_obj)
  File "/usr/src/app/parsing_engines/maccor_engine.py", line 66, in factory
    raise UnsupportedFileTypeError(

Exception Type: UnsupportedFileTypeError at /battDB/exps/add_data/13/
Exception Value: Unknown file extension for Maccor parsing engine '.txt'.

```

**Browser:**
 Chrome - Version 105.0.5195.127 (Official Build) (64-bit)

Aha! Link: https://imperial-college-ict.aha.io/requirements/RSE-132-3

**Comment by dandavies99** [comment]
Update: The biologic part of this issue has been solved in #192 - The data file was all header and no data. There is now a sensible warning to the user if they try and parse a file that looks like this. 

TODO: tackle the maccor part.

**Comment by dandavies99** [comment]
Update from researchers: Maccor cyclers do not output xls files. Maccor files can either be saved as CSV or as a Tab delimited text file. In an ideal world, both csv and txt should be parseable because these are the two file formats a Maccor user could end up with.

This means that the current Maccor parser - that is only capable of parsing xls(x) files - is only useful if a user opens a csv/txt file and re-saves it in Excel. We need to implement txt and csv parsing directly for Maccor.
